src/Diagnostics/Debug/StateCheck.jl

```diff
@@ -0,0 +1,556 @@
+    ###
+    curStats_dict = Dict();
+    curStats_flat = [];
+
+    ######
+    # Create the callback
+    ######
+    cb = EveryXSimulationSteps(ntFreq) do (s = false)
+        # Track which timestep this is
+        nCbCalls = nCbCalls + 1
+        nStep = (nCbCalls - 1) * ntFreq + 1
+        nSStr = @sprintf("%7.7d", nStep - 1)
+
+        ## Free previous curStats_flat if there is one 
+        #  (interesting piece of code - it works, but I am not sure its very elegant!)
+        for i in range(1, length = length(curStats_flat))
+            pop!(curStats_flat)
+        end
+
+
+        ## Print header
+        nprec = min(max(1, prec), 20)
+        if MPI.Comm_rank(MPI.COMM_WORLD) == 0
+            println("# SC +++++++++++ClimateMachine StateCheck call-back start+++++++++++++++++")
+            println("# SC  Step  |   Label    |  Field   |                            Stats                       ")
+        end
+        hVarFmt = "%" * sprintf1("%d", nprec + 8) * "s"
+        minStr = sprintf1(hVarFmt, " min() ")
+        maxStr = sprintf1(hVarFmt, " max() ")
+        aveStr = sprintf1(hVarFmt, " mean() ")
+        stdStr = sprintf1(hVarFmt, " std() ")
+        if MPI.Comm_rank(MPI.COMM_WORLD) == 0
+            println(
+                "# SC =======|============|==========|",
+                minStr,
+                "|",
+                maxStr,
+                "|",
+                aveStr,
+                "|",
+                stdStr,
+                "|",
+            )
+        end
+
+        ## Iterate over the set of MPIStateArrays for this callback
+        for f in fields
+            olabel = f[2]
+            olStr = @sprintf("%12.12s", olabel)
+            mArray = f[1]
+
+            # Get descriptor for MPIStateArray
+
+            V = typeof(mArray).parameters[2]
+
+            ## Iterate over fields in each MPIStateArray
+            #  (use ivar to index individual arrays within the MPIStateArray)
+            ivar = 0
+            statsVal_dict = Dict()
+            for i in 1:length(V.names)
+                for n in flattenednames(fieldtype(V, i), prefix = fieldname(V, i))
+                    ivar = ivar + 1
+                    nStr = @sprintf("%9.9s", n)
+                    if MPI.Comm_rank(MPI.COMM_WORLD) == 0
+                        print("# SC ", nSStr, "|", olStr, "|", nStr, " |")
+                    end
+                    statsString = scstats(mArray, ivar, nprec)
+                    if MPI.Comm_rank(MPI.COMM_WORLD) == 0
+                        println(
+                            statsString[1],
+                            "|",
+                            statsString[2],
+                            "|",
+                            statsString[3],
+                            "|",
+                            statsString[4],
+                            "|",
+                        )
+                    end
+                    statsVal_dict[n] = statsString[5]
+                    append!(
```

Comment:
Shouldn't we know the size of `curStats_flat` beforehand? It would be better to allocate an array before-hand and avoid dynamic memory allocation.